pkg/operators/klusterlet/controllers/bootstrapcontroller/bootstrapcontroller.go

```diff
@@ -0,0 +1,176 @@
+package bootstrapcontroller
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	operatorinformer "github.com/open-cluster-management/api/client/operator/informers/externalversions/operator/v1"
+	operatorlister "github.com/open-cluster-management/api/client/operator/listers/operator/v1"
+	operatorapiv1 "github.com/open-cluster-management/api/operator/v1"
+	"github.com/open-cluster-management/registration-operator/pkg/helpers"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	coreinformer "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+)
+
+// bootstrapController watches bootstrap hub kubeconfig secret, if the secret is changed with hub kube-apiserver ca or apiserver
+// endpoints, this controller will make the klusterlet re-bootstrap to get the new hub kubeconfig from hub cluster by deleting
+// the current hub kubeconfig secret and restart the klusterlet agents
+type bootstrapController struct {
+	kubeClient       kubernetes.Interface
+	klusterletLister operatorlister.KlusterletLister
+	secretLister     corelister.SecretLister
+}
+
+// NewBootstrapController returns a bootstrapController
+func NewBootstrapController(
+	kubeClient kubernetes.Interface,
+	klusterletInformer operatorinformer.KlusterletInformer,
+	secretInformer coreinformer.SecretInformer,
+	recorder events.Recorder) factory.Controller {
+	controller := &bootstrapController{
+		kubeClient:       kubeClient,
+		klusterletLister: klusterletInformer.Lister(),
+		secretLister:     secretInformer.Lister(),
+	}
+	return factory.New().WithSync(controller.sync).
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, err := meta.Accessor(obj)
+			if err != nil {
+				return ""
+			}
+			if accessor.GetName() == helpers.BootstrapHubKubeConfig {
+				return accessor.GetName()
+			}
+			return ""
+		}, secretInformer.Informer()).
+		ToController("BootstrapController", recorder)
+}
+
+func (k *bootstrapController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	bootstrapSecretName := controllerContext.QueueKey()
+	if bootstrapSecretName == "" {
```

Comment:
should also know  the namespace of  bootstrap, and we can find the  related klusterlet by  namespace.

Reply:
ok, I will get the namespace